### PR TITLE
[FLINK-21193][docs] Migrate Flink docs from Jekyll to Hugo 

### PR DESCRIPTION
**Pull request by sjwiesman**
## What is the purpose of the change

Migrates Flinks documentation from Jekyll to Hugo. The documentation can be served locally by following the instructions in the README.

## Brief change log

 I tried to structure this to be as reviewable as possible. 

d56b251  is the actual migration. This was mostly automated with some hand tweaks to get everything to render correctly. I would not attempt to look at this diff :) 

16a718d updates the rest documentation generator to use different css classes so the tables properly render in Hugo. 

bfe9602  update the licensing for the new libraries. I have never modified dependencies in Flink before so please review this commit carefully. 

10b2499 update the documentation README

## Verifying this change

I would recommend building the site locally and clicking through vs looking at the diff. I made an explicit decision not to modify content so we are only concerned with styling. 

**IMPORTANT** Jekyll and Hugo both use a directory called `docs/content` for different purposes. Jekyll uses it to cache build artifacts and as such it has been part of flinks `.gitignore`. This means the Jekyll content will **not** be removed when you check out this branch. Please manually delete these files or the site will fail to build. I strongly encourage you to delete `docs/content` before checking out this branch. 

I will open a PR to flink-web updating the docs contribution page shortly.

### CODE REVIEW

**Comment by sjwiesman** [comment]
cc @rmetzger

**Comment by sjwiesman** [comment]
@zentol Could you please help review 16a718d

**Comment by zentol** [comment]
It seems to be working, but there are some things that feel off:
- the scaling seems to have changed. Pictures are being enlarged while text has become smaller
- sidebar becomes inaccessible when not enough horizontal space is available
- "edit this page" links on the chinese version points to wrong files (I'm not sure whether this link should be visible in the first place)
- the animation when expanding sidebar items is gone, which is unfortunate
- whitespace between sections is too large; it's equivalent to 4 lines for me
- alternating table row background colors looks a bit jank on the metrics page
- hovering over an item in the sidebar can enlarge it such that a line-break occurs

Could you clarify what is bundled in the source release that requires the added licenses? Those changes are likely contained in the monster commit that I couldn't quite make sense of yet.

File: pom.xml [hunk 1]
```diff
@@ -1406,19 +1406,21 @@ under the License.
...
 						<exclude>**/resources/**/jquery*</exclude>
 						<exclude>**/resources/**/bootstrap*</exclude>
 						<exclude>docs/static/flink-header-logo.svg</exclude>
 						<exclude>docs/static/figs/*.svg</exclude>
 						<exclude>docs/resources/**</exclude>
 						<exclude>docs/public/**</exclude>
-						<exclude>docs/static/css/chroma.css</exclude>
+						<exclude>docs/assets/github.css</exclude>
+						<exclude>docs/static/flink-header-logo.svg</exclude>
+						<exclude>docs/static/figs/*.svg</exclude>
```
Comment (zentol) [inline pom.xml:1423]: There are already exclusions for these 2

File: pom.xml [hunk 2]
```diff
@@ -1406,19 +1406,21 @@ under the License.
 						<exclude>**/*.prefs</exclude>
 						<exclude>**/*.log</exclude>
 						<!-- External web libraries. -->
+						<exclude>docs/static/font-awesome/**</exclude>
```
Comment (zentol) [inline pom.xml:1409]: let's move this down to somewhat group things alphabetically

**Comment by zentol** [comment]
There's a lot of wasted space here:

ex

**Comment by zentol** [comment]
In any case, the important thing is that it does work; I think all the things I listed should be fixable in a follow-up. There wasn't any fundamental issue I ran into.

Is the buildbot infrastructure already setup?

What's the plan for 1.11/1.12? Will they remain on the old infrastructure?

**Comment by zentol** [comment]
I assume the `docs/content` directory cannot be changed to something else?

**Comment by zentol** [comment]
`tools/releasing/create_source_release.sh` needs to be updated; it currently excludes the `docs/content` directory.

**Comment by sjwiesman** [comment]
@zentol 

* Updated dependencies:

`LICENSE.book` is for the underlying hugo theme the Flink docs are based on. `LICENSE.katex` and `LICENSE.mermaid` are transitive dependencies of that change. 

`LICENSE.chroma` is the one I am most unsure about. The file `github.css` is used to style our syntax highlighting. It was generated using the command `hugo gen chromastyles --style=github > github.css` where chroma is MIT licensed. My understanding is that the output of this work is derivative and as such `github.css` also needs to be MIT licensed but please correct me if I am mistaken. 

* Config and Rest API styling.

I'm happy to make this look however you feel is best.

* `docs/content` 

Unfortunately we cannot change this. I don't have a satisfying answer here. 

* release-1.11 and release-1.12

The current plan is to leave those as Jekyll based. I estimate it would require a week's worth of work for each branch we need to update. 

* Infra

The build bot is ready to go.

**Comment by alpinegizmo** [comment]
There are ~14 links on the home page that are either missing, or are no longer links. Such as "Hands-on Training" in the Learn Flink paragraph, the 2nd column of links under Explore Flink, and the second paragraph of the Upgrade Flink section.

The "unreleased version" warning at the top of page says "We recommend you use the latest stable version version."

The main font is exceedingly small. Perhaps that's because we're losing so much horizontal space to the TOC in the 3rd column? I think the layout needs some refinement / redesign; the overall feeling I get is "clutter".

The sidebar can be annoying because when you hover over something it switched to bold, when tends to then re-layout the sidebar (at least when magnified to 150%, which is where I can comfortably read the main text), at which point everything jumps around just as I'm about to click.

The search box looks funny:

image

And wow!

**Comment by zentol** [comment]
> the underlying hugo theme the Flink docs are based on
is this actually bundled in the source though, or just injected during the build process + git submodule thingie?

**Comment by sjwiesman** [comment]
@alpinegizmo what browser are you using? The search box does not look like that for me.

**Comment by alpinegizmo** [comment]
@sjwiesman Chrome. The search box only looks that way when I click into it.

**Comment by sjwiesman** [comment]
@zentol It's just during the build because of the sub-module. We don't actually use these libraries. 

@alpinegizmo I've fixed the spacing, font size, and search bar. You can remove the "bold on hover" by commenting out lines 88-99 of `_custom.scss`. I'm not sure if I like that more or not, try it locally and let me know what you think.

**Comment by zentol** [comment]
> It's just during the build because of the sub-module. We don't actually use these libraries.

Then we don't need to bundle the book license.

**Comment by alpinegizmo** [comment]
New font size is very helpful. Also alleviates the bold sidebar issue as well, since the sidebar font is no longer bigger than it needs to be.

In the old sidebar, when you open/close a section, it smoothly animates the opening/closing. The new sidebar simply pops open/closed, which is unfortunate (I find it hard to understand exactly what happened; it's disorienting).

**Comment by sjwiesman** [comment]
> In the old sidebar, when you open/close a section, it smoothly animates the opening/closing. The new sidebar simply pops open/closed, which is unfortunate (I find it hard to understand exactly what happened; it's disorienting).

I agree. I have yet to figure out how we actually accomplished this. I need to dig some more to understand how to replicate.

**Comment by alpinegizmo** [comment]
> I agree. I have yet to figure out how we actually accomplished this. I need to dig some more to understand how to replicate.

That's some bootstrap voodoo. It's done with javascript.

**Comment by alpinegizmo** [comment]
There's a figure missing near the top of http://localhost:1313/docs/deployment/overview/ because the src attribute is missing the initial / character -- it should be "/fig/..."

```
{{< img class="img-fluid" style="margin: 15px" width="100%" src="fig/deployment_overview.svg" alt="Figure for Overview and Reference Architecture" >}} 
```

**Comment by sjwiesman** [comment]
@zentol I've updated the styling for the rest API and configurations. Additonally, I've externalized the css to ` so if we want to continue to tweak it that's where you should look.

File: flink-docs/src/main/java/org/apache/flink/docs/rest/RestAPIDocGenerator.java [hunk 3]
```diff
@@ -220,32 +220,42 @@ private static String createHtmlEntry(MessageHeaders<?, ?, ?> spec) {
...
-            sb.append(
-                    "        <button data-toggle=\"collapse\" data-target=\"#"
-                            + reqHash
-                            + "\">Request</button>\n");
-            sb.append("        <div id=\"" + reqHash + "\" class=\"collapse\">\n");
+            sb.append("      <div class=\"book-expand\">\n");
+            sb.append("        <label>\n");
+            sb.append("          <div class=\"book-expand-head flex justify-between\">\n");
+            sb.append("            <span>Request</span>\n");
+            sb.append("            <span>...</span>\n");
```
Comment (zentol) [inline flink-docs/src/main/java/org/apache/flink/docs/rest/RestAPIDocGenerator.java:227]: How about:

`sb.append("            <span>▾</span>\n");`

**Comment by zentol** [comment]
I think the code tabs have lost a bit of clarity; it's not immediately obvious that you can click them in the first place.

tabs

In a similar vein, the REST API request/response drop-down elements are also not rendered in a way that indicates them to be interactive. The previous approach was admittedly crude. A down-arrow after the text would likely do the trick already.

This will remove some excessive whitespace, and ensure that the url font size isn't smaller than usual text:

```
.rest-api h5 {
    margin-top: .5em;
    margin-bottom: .5em;
    font-size: 1em;
}
```

**Comment by zentol** [comment]
The issue with headings having massive whitespace above them is due to `margin-top` and `padding-top` being set. Either would do the trick I think. (throwing both out isn't so bad either actually...)

**Comment by zentol** [comment]
When some item in the sidebar is expanded such that the scrollbar appears everything is getting resized. Ideally it should just appear without things shifting around.

File: flink-docs/src/main/java/org/apache/flink/docs/rest/RestAPIDocGenerator.java (hunk 3 above)
Comment (zentol) [reply flink-docs/src/main/java/org/apache/flink/docs/rest/RestAPIDocGenerator.java:227]: actually, why not just a single `span` without the `justify-between`? Having the dots/arrow not jump to the right when opened seems nicer?

**Comment by sjwiesman** [comment]
I've fixed the whitespace and added an arrow to the rest API docs. I'm still struggling with the side bar, I'll keep working on it tomorrow.

**Comment by zentol** [comment]
Do we still use bootstrap / jquery for the docs? If not then the bootstrap/anchorjs/jquery could be removed.

**Comment by zentol** [comment]
The python tests are failing because they rely _somehow_ on the documentation ...

```
/__w/3/s/flink-python/dev/.conda/bin/sphinx-build -b html -d _build/doctrees  -a -W . _build/html
Running Sphinx v2.4.4
making output directory... done

Warning, treated as error:
logo file '../../docs/page/img/navbar-brand-logo.jpg' does not exist
Makefile:76: recipe for target 'html' failed
make: *** [html] Error 2

```

As is the docs 404 test which checks for dead links. IOW, `tools/ci/docs.sh` needs to be updated to use hugo. Does hugo automatically catch such errors?

File: tools/releasing/create_source_release.sh [hunk 4]
```diff
@@ -65,7 +65,7 @@ rsync -a \
   --exclude ".git" --exclude ".gitignore" --exclude ".gitattributes" --exclude "azure-pipelines.yml" --exclude ".asf.yaml" \
   --exclude "CHANGELOG" --exclude ".github" --exclude "target" \
   --exclude ".idea" --exclude "*.iml" --exclude ".DS_Store" --exclude "build-target" \
-  --exclude "docs/content" --exclude ".rubydeps" \
+  --exclude "docs/public" --exculde "docs/resources" --exclude "docs/themes" \
```
Comment (zentol) [inline tools/releasing/create_source_release.sh:68]: typo: exculde -> exclude

**Comment by sjwiesman** [comment]
Hugo checks for links, I need to update the docs test to just run the hugo build command.

**Comment by rmetzger** [comment]
Wow, this is amazing! Thanks a lot for your efforts!
I'll take a look at some pages I recently worked on.

File: docs/content/docs/deployment/resource-providers/standalone/docker.md [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+
+Flink introduced `jemalloc` as default memory allocator to resolve memory fragmentation problem (please refer to [FLINK-19125](https://issues.apache.org/jira/browse/FLINK-19125)).
+
+You could switch back to use `glibc` as the memory allocator to restore the old behavior or if any unexpected memory consumption or problem observed
+(and please report the issue via JIRA or mailing list if you found any), by setting environment variable `DISABLE_JEMALLOC` as true:
+
+```sh
+    $ docker run \
+      --env DISABLE_JEMALLOC=true \
+      flink:{{< stable >}}{{< version >}}-scala{{< scala_version >}}{{< /stable >}}{{< unstable >}}
```
Comment (rmetzger) [inline docs/content/docs/deployment/resource-providers/standalone/docker.md:349]:
```suggestion
      flink:{{< stable >}}{{< version >}}-scala{{< scala_version >}}{{< /stable >}}{{< unstable >}}latest{{< /unstable >}} <jobmanager|standalone-job|taskmanager>
```

**Comment by rmetzger** [comment]
The configuration keys anchors don't seem to work. http://localhost:1313/docs/deployment/config/#jobmanager-rpc-address shows the top of the page. Do I need to rebuild the includes?

**Comment by rmetzger** [comment]
There's at least 2 broken tables:
image

**Comment by sjwiesman** [comment]
@rmetzger thanks for the review. I've fixed the anchor tags for the configuration and the tables. I'm finishing up the licensing right now and will then rebase and squash my fixup commits to get fix the merge conflict of the updated configurations.

File: tools/ci/docs.sh [hunk 6]
```diff
@@ -17,34 +17,20 @@
...
+tar -zxvf $HUGO_ARTIFACT
 
+hugo --source docs
 
-for i in `seq 1 90`;
-do
-	echo "Waiting for server..."
-	curl -Is http://localhost:4000 --fail
-	if [ $? -eq 0 ]; then
-		./docs/check_links.sh
```
Comment (zentol) [inline tools/ci/docs.sh:42]: I suppose we can delete this file now

Comment (sjwiesman) [reply tools/ci/docs.sh:42]: +1

**Comment by alpinegizmo** [comment]
I recommend making this change:

```
diff --git a/docs/assets/_custom.scss b/docs/assets/_custom.scss
index f9641bdff3..dfacac7a9b 100644
--- a/docs/assets/_custom.scss
+++ b/docs/assets/_custom.scss
@@ -77,6 +77,10 @@ aside nav ul {
   }
 }
 
+.book-page {
+    padding: 2rem 2rem;
+}
+
 .book-search {
        border: 2px solid #ebebeb;
 }
```

This increases the left and right padding around the main part of the page, making it more like the current docs. Also looks better, imo.

**Comment by sjwiesman** [comment]
I've rebased and squashed my fixup commits. I will keep monitoring CI to ensure everything passes but if there are not any other issues people view as absolute blockers I would like to get this merged in and address any further comments as follow-up tickets to unblock other documentation PR's and avoid constant rebasing.

**Comment by zentol** [comment]
If CI passes then I'm OK with merging things as is.

**Review by rmetzger — APPROVED** [review]
Thanks a lot for this effort!

**Comment by rmetzger** [comment]
No need to address this right now, but the 1.11 release notes have some issues:

image

**Pull request closed without merging**